### PR TITLE
Override MetadataName for PEParameterSymbol to return actual name from metadata.

### LINKED ISSUE #8088: Behavior change around embedding unnamed parameters (NoPia)
While embedding parameters that don't have names assigned in metadata, Roslyn compilers emit them with default name ("value" for C#, "Param" for VB). Methods that have more than one parameter (for example, delegate constructors) could end up with duplicate parameter names in metadata. This is not a violation of ECMA-335, Section II.22.33 "Param : 0x08" doesn't have a requirement for parameter names to be unique. The resulting assembly is verifiable and can be successfully disassembled with ILDAM and reassembled back with ILASM. With one wrinkle though, ILASM reports a warning about a duplicate parameter name. It is possible that other tools out there could be affected too.

VS2013 C# compiler doesn't assign names to unnamed parameters in the process of embedding them. VB is probably doing the same.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #8088.

Behavior change can be observed through ISymbol.MetadataName property.

@dotnet/roslyn-compiler Please review.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs [hunk 1]
```diff
@@ -304,6 +314,30 @@ public override string Name
...
+                    try
+                    {
+                        return _moduleSymbol.Module.GetParamNameOrThrow(_handle);
+                    }
+                    catch (BadImageFormatException)
+                    {
+                    }
+                }
+
+                return String.Empty;
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: string.Empty

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs [hunk 2]
```diff
@@ -304,6 +314,30 @@ public override string Name
             }
         }
 
+        public override string MetadataName
+        {
+            get
+            {   
+                if (_name.Length != s_defaultNameLength || _name != s_defaultName)
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: This seems like a rather indirect way of checking for the `null` name case and causes us to do extra work when the name is `value` (which is pretty common).  Way not add an extra value in packed flags?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: Sure, we can do that.

**Comment by AlekseyTs** [comment]
The feedback has been addressed, @dotnet/roslyn-compiler please review.

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs [hunk 3]
```diff
@@ -304,6 +320,14 @@ public override string Name
             }
         }
 
+        public override string MetadataName
+        {
+            get
+            {   
+                return HasNameInMetadata ? _name : string.Empty;
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: I was under the impression that empty string was an invalid metadata name for parameters (but `null` was legal).  It looks like this gets translated under the hood to `StringIdx(0)`:

https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/PEWriter/MetadataHeapsBuilder.cs#L237

I can't tell if that gets translated to an empty string or `null` though.

File: src/Compilers/Test/Resources/Core/SymbolsTests/NoPia/ParametersWithoutNames.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Reflection;
+using System.Reflection.Emit;
+using System.Runtime.InteropServices;
+
+
+class Program
```
Comment (jaredpar) [inline src/Compilers/Test/Resources/Core/SymbolsTests/NoPia/ParametersWithoutNames.cs]: Accessiility modifier

File: src/Compilers/Test/Resources/Core/SymbolsTests/NoPia/ParametersWithoutNames.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+{
+    /// <summary>
+    /// Compile and run this program to generate ParametersWithoutNames.dll
+    /// </summary>
+    static void Main()
```
Comment (jaredpar) [inline src/Compilers/Test/Resources/Core/SymbolsTests/NoPia/ParametersWithoutNames.cs]: accessibility modifier

**Comment by cston** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs (hunk 3 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: Yes, it is possible to get confused with "null" terminology. I got confused myself. Index 0 is in fact what ECMA spec calls null, and it turns out, when we import/emit strings values from/to metadata, it is equivalent to an empty string.

File: src/Compilers/Test/Resources/Core/SymbolsTests/NoPia/ParametersWithoutNames.cs (hunk 4 above)
Comment (AlekseyTs) [inline src/Compilers/Test/Resources/Core/SymbolsTests/NoPia/ParametersWithoutNames.cs]: This file isn't compiled, it is there for informational purposes only. But, if you insist, I can add explicit accessibility modifier.

File: src/Compilers/Test/Resources/Core/SymbolsTests/NoPia/ParametersWithoutNames.cs (hunk 5 above)
Comment (AlekseyTs) [inline src/Compilers/Test/Resources/Core/SymbolsTests/NoPia/ParametersWithoutNames.cs]: Same as above.

**Comment by jaredpar** [comment]
:+1:

**Pull request merged**